File: articles/probability-variation-biological-inference/python/bootstrap_uncertainty.py
"""
Bootstrap uncertainty workflow.

Run:
    python python/bootstrap_uncertainty.py
"""

from pathlib import Path

import pandas as pd

from probability_core import bootstrap_mean


ARTICLE_DIR = Path(__file__).resolve().parents[1]
MEASURE_PATH = ARTICLE_DIR / "data" / "biological_measurements.csv"


def main() -> None:
    data = pd.read_csv(MEASURE_PATH)

    rows = []

    for group, group_df in data.groupby("group"):
        result = bootstrap_mean(group_df["value"], n_bootstrap=5000, seed=42)

        rows.append(
            {
                "group": group,
                "n": len(group_df),
                **result,
            }
        )

    print(pd.DataFrame(rows).round(5).to_string(index=False))


if __name__ == "__main__":
    main()

File: articles/probability-variation-biological-inference/r/bootstrap_uncertainty.R
# Bootstrap uncertainty workflow in R.

measure_path <- file.path("data", "biological_measurements.csv")
if (!file.exists(measure_path)) {
  measure_path <- file.path("..", "data", "biological_measurements.csv")
}

set.seed(42)

data <- read.csv(measure_path)

bootstrap_mean <- function(x, n_boot = 5000) {
  boot <- replicate(n_boot, mean(sample(x, replace = TRUE)))

  c(
    observed_mean = mean(x),
    bootstrap_mean = mean(boot),
    ci_lower = quantile(boot, 0.025),
    ci_upper = quantile(boot, 0.975)
  )
}

rows <- lapply(split(data$value, data$group), bootstrap_mean)

summary_df <- data.frame(
  group = names(rows),
  do.call(rbind, rows),
  row.names = NULL
)

print(round(summary_df, 5))
